File: apple-client/Hearth/Core/Sources/HearthUI/Persona/PersonaOrb.swift
//
//  PersonaOrb.swift
//  Hearth  (shared by the app target AND the Hearth WidgetExtension target)
//
//  The actual Canvas drawing of Sulivan's sphere_particle persona — a single
//  frame for a given animation time. The app wraps this in a TimelineView for
//  live animation (PersonaCanvasView); widgets render one static frame (they
//  can't run TimelineView(.animation)). Ported from Echo's PersonaCanvas.kt.
//
//  IMPORTANT: tick this file into the Hearth WidgetExtension target membership
//  in Xcode (it is reused by PersonaStatusWidget).
//

import SwiftUI
import HearthCore

public struct PersonaOrb: View {
    public var state: HearthState
    /// TTS amplitude 0..1 (drives the speaking waveform).
    var pulse: Double = 0
    /// Animation time source. Use `.now` while animating; any fixed date when static.
    var date: Date = .distantPast
    /// 0 = orbit, 1 = speaking waveform (the app animates this; widgets pass 0).
    var waveBlend: Double = 0
    var reducedMotion: Bool = false

    /// The persona's orb colours, data-driven from its config. Defaults to the
    /// warm HearthPalette fallback so widgets (which have no live connection)
    /// render warm too; the app injects the live palette via PersonaCanvasView.
    public var palette: PersonaPalette = .fallback

    private let particles = PersonaParticle.field()

    /// Explicit, because a public struct's synthesized memberwise initializer is
    /// internal and would leave this View visible from the app targets but not
    /// constructible by them.
    public init(
        state: HearthState,
        pulse: Double = 0,
        date: Date = .distantPast,
        waveBlend: Double = 0,
        reducedMotion: Bool = false,
        palette: PersonaPalette = .fallback
    ) {
        self.state = state
        self.pulse = pulse
        self.date = date
        self.waveBlend = waveBlend
        self.reducedMotion = reducedMotion
        self.palette = palette
    }

    public var body: some View {
        Canvas { ctx, size in draw(into: ctx, size: size) }
    }

    private func draw(into ctx: GraphicsContext, size: CGSize) {
        let p = pulse.clamped(0, 1)

        let elapsed = reducedMotion ? 0 : date.timeIntervalSinceReferenceDate
        let time = (elapsed.truncatingRemainder(dividingBy: 12) / 12) * 2 * .pi
        let breath = reducedMotion ? 0 : triangle(elapsed, half: 3.8)
        let shimmer = reducedMotion ? 0.5 : triangle(elapsed, half: 1.6)
        let wave = reducedMotion ? 0 : elapsed.truncatingRemainder(dividingBy: 1.3) / 1.3

        let v = visual(for: state, shimmer: shimmer, breath: breath, pulse: p)

        let cx = size.width / 2
        let cy = size.height / 2
        let field = min(size.width, size.height) / 2
        let sphereRadius = field * 0.34 * v.swell
        let orbitRadius = field * 0.92 * v.spread

        // Radial glow halo behind the sphere.
        let haloR = sphereRadius * 2.6
        ctx.fill(
            circlePath(cx: cx, cy: cy, r: haloR),
            with: .radialGradient(
                Gradient(colors: [v.glow.opacity(0.55 * v.brightness), v.glow.opacity(0)]),
                center: CGPoint(x: cx, y: cy), startRadius: 0, endRadius: haloR
            )
        )

        // Particle field (orbit at rest, easing into the speaking waveform).
        let twoPi = 2.0 * Double.pi
        let waveHalfWidth = field * 0.95
        let waveCycles = 2.4
        let travel = wave * twoPi
        let waveAmp = field * (0.05 + 0.32 * p)
        let lastIdx = Double(max(1, PersonaParticle.count - 1))

        for (i, part) in particles.enumerated() {
            let angle = part.baseAngle + time * part.angularSpeed * v.motion
            let r = orbitRadius * part.ringRadius
            let ox = cx + cos(angle) * r
            let oy = cy + sin(angle) * r * part.ringEccentricity
            let depth = sin(angle) * 0.5 + 0.5

            let frac = Double(i) / lastIdx
            let phaseArg = frac * waveCycles * twoPi - travel + part.phase * 0.15
            let wx = cx + (frac - 0.5) * 2 * waveHalfWidth
            let wy = cy + sin(phaseArg) * waveAmp * (0.7 + 0.6 * part.sizeFactor)
                + (part.sizeFactor - 0.5) * field * 0.06

            let px = ox + (wx - ox) * waveBlend
            let py = oy + (wy - oy) * waveBlend

            let twinkle = reducedMotion ? 0.7 : (0.55 + 0.45 * sin(time + part.phase) * 0.5 + 0.225)
            let orbitAlpha = (0.30 + 0.55 * depth) * twinkle
            let crest = 0.5 + 0.5 * sin(phaseArg)
            let waveAlpha = 0.45 + 0.55 * crest * (0.5 + 0.5 * p)
            let alpha = ((orbitAlpha + (waveAlpha - orbitAlpha) * waveBlend) * v.brightness).clamped(0, 1)

            let dotDepth = 0.6 + 0.8 * depth
            let dotR = field * 0.012
                * (dotDepth + (1.0 - dotDepth) * waveBlend)
                * (0.7 + 0.6 * part.sizeFactor)

            ctx.fill(circlePath(cx: px, cy: py, r: dotR), with: .color(palette.particleColor.opacity(alpha)))
        }

        // The sphere: bright core fading to its edge tint at the rim.
        ctx.fill(
            circlePath(cx: cx, cy: cy, r: sphereRadius),
            with: .radialGradient(
                Gradient(stops: [
                    .init(color: v.core.opacity((0.92 * v.brightness).clamped(0, 1)), location: 0),
                    .init(color: v.core.opacity((0.65 * v.brightness).clamped(0, 1)), location: 0.55),
                    .init(color: v.edge.opacity((0.45 * v.brightness).clamped(0, 1)), location: 1),
                ]),
                center: CGPoint(x: cx, y: cy), startRadius: 0, endRadius: sphereRadius
            )
        )

        // Edge ring — redundant state cue (the listening accent while LISTENING,
        // the idle accent otherwise; both from the persona palette).
        ctx.stroke(
            Path(ellipseIn: CGRect(x: cx - sphereRadius, y: cy - sphereRadius,
                                   width: sphereRadius * 2, height: sphereRadius * 2)),
            with: .color(v.edge.opacity((0.55 * v.brightness).clamped(0, 1))),
            lineWidth: field * 0.012
        )
    }

    // MARK: - State visuals

    private struct Visual {
        var core: Color, glow: Color, edge: Color
        var brightness: Double, swell: Double, spread: Double, motion: Double
    }

    private func visual(for state: HearthState, shimmer: Double, breath: Double, pulse: Double) -> Visual {
        // Colours are the persona's palette (data-driven); the brightness/swell/
        // spread/motion choreography is unchanged. `core` is the bead body, `glow`
        // the halo, `edge` the redundant state-cue ring.
        let core = palette.sphereColor
        let idle = palette.idleColor
        switch state {
        case .LOADING, .IDLE:
            return Visual(core: core, glow: idle, edge: idle,
                          brightness: 0.45 + 0.10 * breath, swell: 1.0 + 0.03 * breath,
                          spread: 1.0, motion: 0.35)
        case .LISTENING:
            return Visual(core: core, glow: idle, edge: palette.listeningColor,
                          brightness: 0.95, swell: 1.12, spread: 1.22, motion: 1.0)
        case .THINKING:
            let thinking = palette.thinkingColor
            return Visual(core: blend(core, thinking, 0.25 * shimmer),
                          glow: blend(idle, thinking, shimmer),
                          edge: blend(idle, thinking, shimmer),
                          brightness: 0.7 + 0.15 * shimmer,
                          swell: 1.0 + 0.04 * sin(shimmer * .pi),
                          spread: 1.05, motion: 0.8)
        case .SPEAKING:
            let speaking = palette.speakingColor
            return Visual(core: blend(core, speaking, 0.2),
                          glow: blend(idle, speaking, 0.4 + 0.4 * pulse),
                          edge: blend(idle, speaking, 0.5),
                          brightness: (0.75 + 0.25 * pulse).clamped(0, 1),
                          swell: 1.0 + 0.18 * pulse, spread: 1.0 + 0.15 * pulse, motion: 1.0)
        }
    }

    // MARK: - Helpers

    private func circlePath(cx: CGFloat, cy: CGFloat, r: CGFloat) -> Path {
        Path(ellipseIn: CGRect(x: cx - r, y: cy - r, width: r * 2, height: r * 2))
    }

    private func triangle(_ elapsed: Double, half: Double) -> Double {
        let x = elapsed.truncatingRemainder(dividingBy: 2 * half) / half
        return x <= 1 ? x : 2 - x
    }

    private func blend(_ a: Color, _ b: Color, _ t: Double) -> Color {
        let tt = t.clamped(0, 1)
        let ca = a.rgba, cb = b.rgba
        return Color(red: ca.r + (cb.r - ca.r) * tt,
                     green: ca.g + (cb.g - ca.g) * tt,
                     blue: ca.b + (cb.b - ca.b) * tt)
    }
}

// MARK: - Particle field

struct PersonaParticle {
    let baseAngle: Double
    let ringRadius: Double
    let ringEccentricity: Double
    let angularSpeed: Double
    let sizeFactor: Double
    let phase: Double

    static let count = 96

    /// Deterministic LCG-seeded field (same look every launch), matching Echo.
    static func field() -> [PersonaParticle] {
        var seed = Int32(bitPattern: 0x9E37_79B9)
        func next() -> Double {
            seed = seed &* 1103515245 &+ 12345
            return Double((UInt32(bitPattern: seed) >> 8) & 0xFFFF) / 65535.0
        }
        let twoPi = 2 * Double.pi
        return (0..<count).map { _ in
            PersonaParticle(
                baseAngle: next() * twoPi,
                ringRadius: 0.45 + next() * 0.55,
                ringEccentricity: 0.55 + next() * 0.35,
                angularSpeed: (0.4 + next() * 0.9) * (next() > 0.5 ? 1 : -1),
                sizeFactor: 0.4 + next() * 0.6,
                phase: next() * twoPi
            )
        }
    }
}

// MARK: - Small numeric / colour helpers

extension Double {
    func clamped(_ lo: Double, _ hi: Double) -> Double { Swift.min(Swift.max(self, lo), hi) }
}

// `Color.rgba` used to live here. It moved to HearthCore's ColorHex.swift when
// the package split, because HearthPalette.mixed(with:amount:) needs it and
// HearthCore cannot reach up into HearthUI.

File: android-client/app/src/main/java/com/hearth/app/ui/persona/PersonaOrb.kt
package com.hearth.app.ui.persona

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.withFrameMillis
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import com.hearth.core.models.HearthState
import com.hearth.core.persona.PersonaParticle
import com.hearth.core.persona.PersonaPalette
import com.hearth.core.persona.Rgb
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

/**
 * The `sphere_particle` persona: a bead in a field of orbiting dots. Ported
 * from the iOS `PersonaOrb` and `PersonaCanvasView`, which came from Echo's
 * `PersonaCanvas.kt`.
 *
 * THIS IS THE DEFAULT APPEARANCE, and it is what everything else falls back
 * to: a model persona whose asset has not arrived, a face whose geometry has
 * not, a persona the house knows nothing about yet. Android's fallback used to
 * be a flat filled circle, which is indistinguishable from a rendering failure.
 *
 * The field is choreographed rather than simulated, which is what lets it ease
 * from an orbit into a speaking waveform: at rest the dots ride elliptical
 * rings, and while SPEAKING they lay themselves out left to right along a sine
 * whose amplitude is the voice. The blend between the two is animated over a
 * third of a second so the change reads as a transition rather than a cut.
 */
@Composable
fun PersonaOrb(
    state: HearthState,
    /** TTS amplitude 0..1; drives the speaking waveform. */
    pulse: Float = 0f,
    palette: PersonaPalette = PersonaPalette.fallback,
    reduceMotion: Boolean = false,
    modifier: Modifier = Modifier,
) {
    // Same frame clock as the flame, for the same reason: a Canvas does not
    // animate itself.
    val clock = remember { OrbClock() }
    var frameTick by remember { mutableIntStateOf(0) }

    LaunchedEffect(reduceMotion) {
        if (reduceMotion) {
            clock.elapsed = 0.0
            frameTick++
            return@LaunchedEffect
        }
        while (true) {
            withFrameMillis { ms ->
                if (clock.origin == 0L) clock.origin = ms
                clock.elapsed = (ms - clock.origin) / 1000.0
                frameTick++
            }
        }
    }

    // The one value that is NOT on the frame clock: how far the field has
    // travelled from orbit toward waveform. It is a transition between two
    // layouts rather than a continuous motion, so it gets a real animation.
    val waveBlend by animateFloatAsState(
        targetValue = if (state == HearthState.SPEAKING && !reduceMotion) 1f else 0f,
        animationSpec = tween(durationMillis = 320),
        label = "waveBlend",
    )

    val field = remember { PersonaParticle.field() }
    val inputs = remember { OrbInputs() }
    inputs.state = state
    inputs.pulse = pulse.toDouble()
    inputs.reduceMotion = reduceMotion
    inputs.palette = palette
    inputs.waveBlend = waveBlend.toDouble()

    Canvas(modifier = modifier) {
        @Suppress("UNUSED_EXPRESSION")
        frameTick
        drawOrb(clock.elapsed, field, inputs)
    }
}

private class OrbClock {
    var origin: Long = 0
    var elapsed: Double = 0.0
}

private class OrbInputs {
    var state: HearthState = HearthState.IDLE
    var pulse: Double = 0.0
    var reduceMotion: Boolean = false
    var waveBlend: Double = 0.0
    var palette: PersonaPalette = PersonaPalette.fallback
}

/** What a turn state does to the bead. Colours are the persona's; the
 *  choreography is the same on every client. */
private class Visual(
    val core: Rgb,
    val glow: Rgb,
    val edge: Rgb,
    val brightness: Double,
    val swell: Double,
    val spread: Double,
    val motion: Double,
)

private fun DrawScope.drawOrb(
    elapsed: Double,
    particles: List<PersonaParticle>,
    inputs: OrbInputs,
) {
    val p = inputs.pulse.coerceIn(0.0, 1.0)
    val still = inputs.reduceMotion
    val t = if (still) 0.0 else elapsed

    val time = (t.mod(12.0) / 12) * 2 * PI
    val breath = if (still) 0.0 else triangle(t, 3.8)
    val shimmer = if (still) 0.5 else triangle(t, 1.6)
    val wave = if (still) 0.0 else t.mod(1.3) / 1.3

    val v = visual(inputs.state, inputs.palette, shimmer, breath, p)

    val cx = size.width / 2.0
    val cy = size.height / 2.0
    val field = minOf(size.width, size.height) / 2.0
    val sphereRadius = field * 0.34 * v.swell
    val orbitRadius = field * 0.92 * v.spread

    // The halo behind the bead.
    val haloR = sphereRadius * 2.6
    val center = Offset(cx.toFloat(), cy.toFloat())
    drawCircle(
        brush = Brush.radialGradient(
            colors = listOf(
                v.glow.toColor((0.55 * v.brightness).toFloat()),
                v.glow.toColor(0f),
            ),
            center = center,
            radius = haloR.toFloat(),
        ),
        radius = haloR.toFloat(),
        center = center,
    )

    // The field: an orbit at rest, easing into the speaking waveform.
    val twoPi = 2 * PI
    val waveHalfWidth = field * 0.95
    val waveCycles = 2.4
    val travel = wave * twoPi
    val waveAmp = field * (0.05 + 0.32 * p)
    val lastIdx = max(1, PersonaParticle.COUNT - 1).toDouble()
    val blend = inputs.waveBlend

    particles.forEachIndexed { i, part ->
        val angle = part.baseAngle + time * part.angularSpeed * v.motion
        val r = orbitRadius * part.ringRadius
        val ox = cx + cos(angle) * r
        val oy = cy + sin(angle) * r * part.ringEccentricity
        // Where on its ellipse the dot is, read as depth: the far half is
        // dimmer and smaller, which is the whole of the 2D bead's roundness.
        val depth = sin(angle) * 0.5 + 0.5

        val frac = i / lastIdx
        val phaseArg = frac * waveCycles * twoPi - travel + part.phase * 0.15
        val wx = cx + (frac - 0.5) * 2 * waveHalfWidth
        val wy = cy + sin(phaseArg) * waveAmp * (0.7 + 0.6 * part.sizeFactor) +
            (part.sizeFactor - 0.5) * field * 0.06

        val px = ox + (wx - ox) * blend
        val py = oy + (wy - oy) * blend

        val twinkle = if (still) 0.7 else 0.55 + 0.45 * sin(time + part.phase) * 0.5 + 0.225
        val orbitAlpha = (0.30 + 0.55 * depth) * twinkle
        val crest = 0.5 + 0.5 * sin(phaseArg)
        val waveAlpha = 0.45 + 0.55 * crest * (0.5 + 0.5 * p)
        val alpha = ((orbitAlpha + (waveAlpha - orbitAlpha) * blend) * v.brightness)
            .coerceIn(0.0, 1.0)

        val dotDepth = 0.6 + 0.8 * depth
        val dotR = field * 0.012 *
            (dotDepth + (1.0 - dotDepth) * blend) *
            (0.7 + 0.6 * part.sizeFactor)

        drawCircle(
            color = inputs.palette.particle.toColor(alpha.toFloat()),
            radius = dotR.toFloat(),
            center = Offset(px.toFloat(), py.toFloat()),
        )
    }

    // The bead: a bright core fading to its edge tint at the rim.
    drawCircle(
        brush = Brush.radialGradient(
            colorStops = arrayOf(
                0.00f to v.core.toColor((0.92 * v.brightness).coerceIn(0.0, 1.0).toFloat()),
                0.55f to v.core.toColor((0.65 * v.brightness).coerceIn(0.0, 1.0).toFloat()),
                1.00f to v.edge.toColor((0.45 * v.brightness).coerceIn(0.0, 1.0).toFloat()),
            ),
            center = center,
            radius = sphereRadius.toFloat(),
        ),
        radius = sphereRadius.toFloat(),
        center = center,
    )

    // The edge ring: a REDUNDANT state cue, so the turn is legible in
    // peripheral vision and to anyone who cannot separate the state colours.
    drawCircle(
        color = v.edge.toColor((0.55 * v.brightness).coerceIn(0.0, 1.0).toFloat()),
        radius = sphereRadius.toFloat(),
        center = center,
        style = Stroke(width = (field * 0.012).toFloat()),
    )
}

private fun visual(
    state: HearthState,
    palette: PersonaPalette,
    shimmer: Double,
    breath: Double,
    pulse: Double,
): Visual {
    val core = palette.sphere
    val idle = palette.idle
    return when (state) {
        HearthState.LOADING, HearthState.IDLE -> Visual(
            core = core, glow = idle, edge = idle,
            brightness = 0.45 + 0.10 * breath,
            swell = 1.0 + 0.03 * breath,
            spread = 1.0, motion = 0.35,
        )

        HearthState.LISTENING -> Visual(
            core = core, glow = idle, edge = palette.listening,
            brightness = 0.95, swell = 1.12, spread = 1.22, motion = 1.0,
        )

        HearthState.THINKING -> Visual(
            core = core.mix(palette.thinking, (0.25 * shimmer).toFloat()),
            glow = idle.mix(palette.thinking, shimmer.toFloat()),
            edge = idle.mix(palette.thinking, shimmer.toFloat()),
            brightness = 0.7 + 0.15 * shimmer,
            swell = 1.0 + 0.04 * sin(shimmer * PI),
            spread = 1.05, motion = 0.8,
        )

        HearthState.SPEAKING -> Visual(
            core = core.mix(palette.speaking, 0.2f),
            glow = idle.mix(palette.speaking, (0.4 + 0.4 * pulse).toFloat()),
            edge = idle.mix(palette.speaking, 0.5f),
            brightness = min(1.0, 0.75 + 0.25 * pulse),
            swell = 1.0 + 0.18 * pulse,
            spread = 1.0 + 0.15 * pulse, motion = 1.0,
        )
    }
}

/** A triangle wave: up over [half] seconds, back down over the next. */
private fun triangle(elapsed: Double, half: Double): Double {
    val x = elapsed.mod(2 * half) / half
    return if (x <= 1) x else 2 - x
}

private fun Rgb.toColor(alpha: Float) = Color(r, g, b, alpha.coerceIn(0f, 1f))
